google/cloud/spanner/numeric.h

```diff
@@ -118,25 +137,195 @@ class Numeric {
 
   /// Relational operators
   ///@{
-  friend bool operator==(Numeric const& a, Numeric const& b) {
+  friend bool operator==(Decimal const& a, Decimal const& b) {
+    // Decimal-value equality, which only depends on the canonical
+    // representation, not the mode. The representation may be "NaN"
+    // in kPostgreSQL mode, but unlike typical NaN implementations,
+    // PostgreSQL considers NaN values as equal, so that they may be
+    // sorted. We do the same.
     return a.rep_ == b.rep_;
   }
-  friend bool operator!=(Numeric const& a, Numeric const& b) {
+  friend bool operator!=(Decimal const& a, Decimal const& b) {
     return !(a == b);
   }
   ///@}
 
-  /// Outputs string representation of the `Numeric` to the provided stream.
-  friend std::ostream& operator<<(std::ostream& os, Numeric const& n) {
-    return os << n.ToString();
+  /// Outputs string representation of the `Decimal` to the provided stream.
+  friend std::ostream& operator<<(std::ostream& os, Decimal const& d) {
+    return os << d.ToString();
   }
 
  private:
-  friend StatusOr<Numeric> spanner_internal::MakeNumeric(std::string s);
-  explicit Numeric(std::string rep) : rep_(std::move(rep)) {}
-  std::string rep_;  // a valid and canonical NUMERIC representation
+  template <DecimalMode Mode>
+  friend StatusOr<Decimal<Mode>> spanner_internal::
+#if defined(__GNUC__) && !defined(__clang__)
+      GOOGLE_CLOUD_CPP_NS::
+#endif
+          MakeDecimal(std::string);
+
+  explicit Decimal(std::string rep) : rep_(std::move(rep)) {}
+
+  std::string rep_;  // a valid and canonical decimal representation
 };
 
+template <DecimalMode Mode>
+constexpr std::size_t Decimal<Mode>::kIntPrec;
+template <DecimalMode Mode>
+constexpr std::size_t Decimal<Mode>::kFracPrec;
+
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_END
+}  // namespace spanner
+
+// Internal implementation details that callers should not use.
+namespace spanner_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+// Like `std::to_string`, but also supports Abseil 128-bit integers.
+template <typename T>
+std::string ToString(T&& value) {
+  return std::to_string(std::forward<T>(value));
+}
+std::string ToString(absl::int128 value);
+std::string ToString(absl::uint128 value);
+
+inline bool IsDigit(char ch) {
+  return std::isdigit(static_cast<unsigned char>(ch)) != 0;
+}
+
+inline bool IsSpace(char ch) {
+  return std::isspace(static_cast<unsigned char>(ch)) != 0;
+}
+
+bool IsNaN(std::string const& rep);
+bool IsCanonical(absl::string_view sign_part, absl::string_view int_part,
+                 absl::string_view frac_part, std::size_t int_prec,
+                 std::size_t frac_prec);
+void Round(std::deque<char>& int_rep, std::deque<char>& frac_rep,
+           std::size_t frac_prec);
+
+Status InvalidArgument(std::string message);
+Status OutOfRange(std::string message);
+Status DataLoss(std::string message);
+
+template <spanner::DecimalMode Mode>
+StatusOr<spanner::Decimal<Mode>> MakeDecimal(std::string s) {
```

Comment:
Can we refactor this function to the `.cc` file?  Maybe:

```cc
StatusOr<std::string> MakeDecimalRep(std::string s, std::size_t int_precision, std::size_t frac_precision);

template <...>
StatusOr<spanner::Decimal<...>> MakeDecimal(std::string s) {
  auto rep = MakeDecimalRep(std::move(s), Decimal<...>::kIntPrecision, Decimal<...>::kFracPrecision);
  if (!rep) return std::move(rep).status();
  return Decimal<...>(std::move(rep));
}
```

Reply:
Yes.  Done.

Reply:
Thanks.